src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,273 @@
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of the length of the cached range
+            var cachedRangeLength = 0;
+
+            var tokens = GetCachedTokens(requestedRange, lineToTokensDict, ref cachedRangeStart, ref cachedRangeLength);
+            if (tokens.Count == 0)
+            {
+                // We couldn't find any tokens associated with the passed-in range
+                cachedTokens = null;
+                return false;
+            }
+
+            Assumes.NotNull(cachedRangeStart);
+
+            // We can potentially return a partial range match if we can't find a full range match.
+            var endPosition = new Position { Line = cachedRangeStart.Line + cachedRangeLength, Character = 0 };
+            var range = new Range { Start = cachedRangeStart, End = endPosition };
+
+            cachedTokens = (range, tokens);
+            return true;
+        }
+
+        private void CacheTokensPerLine(Range range, int[] tokens, Dictionary<int, IReadOnlyList<int>> lineToTokensDict)
+        {
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            var tokenIndex = 0;
+            while (tokenIndex < tokens.Length)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line)
+                {
+                    tokenIndex += TokenSize;
+                    continue;
+                }
+
+                if (absoluteLine > range.End.Line)
+                {
+                    break;
+                }
+
+                lock (_dictLock)
+                {
+                    if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                    {
+                        // This line is already cached, so we can skip it
+                        tokenIndex += TokenSize;
+                        continue;
+                    }
+                }
+
+                // Cache until we hit the next line or the end of the tokens
+                var lineTokens = new List<int>();
+                do
+                {
+                    lineTokens.Add(tokens[tokenIndex]);
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
+
+                    tokenIndex += 5;
+                }
+                while (tokenIndex < tokens.Length && tokens[tokenIndex] == 0);
+
+                lock (_dictLock)
+                {
+                    lineToTokensDict.Add(absoluteLine, lineTokens);
```

Comment:
this is very unsafe and will throw in practic if you have two threads that made it here.  you either need to check and bail in this lock, or just use `lineToTokensDict[absoluteLine] = lineTokens;` to let last thread win.